File: swift-bridge/Sources/AudioToolboxBridge/AudioComponent.swift
import AudioToolbox
import AudioUnit
import CoreAudio
import CoreFoundation
import Foundation

private final class AudioComponentBox {
    let value: AudioComponent

    init(_ value: AudioComponent) {
        self.value = value
    }
}

private final class AudioComponentInstanceBox {
    var value: AudioComponentInstance

    init(_ value: AudioComponentInstance) {
        self.value = value
    }

    deinit {
        AudioComponentInstanceDispose(value)
    }
}

private func audioComponent(from raw: UnsafeMutableRawPointer?) -> AudioComponent {
    castOpaque(raw, to: AudioComponent.self)
}

private func audioComponentInstance(from raw: UnsafeMutableRawPointer?) -> AudioComponentInstance {
    castOpaque(raw, to: AudioComponentInstance.self)
}

@_cdecl("at_audio_component_count")
public func at_audio_component_count(_ description: UnsafePointer<AudioComponentDescription>?) -> UInt32 {
    guard let description else {
        return 0
    }
    return AudioComponentCount(description)
}

@_cdecl("at_audio_component_find_next")
public func at_audio_component_find_next(
    _ previousRaw: UnsafeMutableRawPointer?,
    _ description: UnsafePointer<AudioComponentDescription>?
) -> UnsafeMutableRawPointer? {
    guard let description else {
        return nil
    }
    let previous = previousRaw.map { audioComponent(from: $0) }
    guard let component = AudioComponentFindNext(previous, description) else {
        return nil
    }
    return retainObject(AudioComponentBox(component))
}

@_cdecl("at_audio_component_raw")
public func at_audio_component_raw(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let handle else {
        return nil
    }
    let box: AudioComponentBox = takeUnretained(handle)
    return toRawPointer(box.value)
}

@_cdecl("at_audio_component_release")
public func at_audio_component_release(_ handle: UnsafeMutableRawPointer?) {
    guard let handle else {
        return
    }
    releaseObject(handle, as: AudioComponentBox.self)
}

@_cdecl("at_audio_component_copy_name")
public func at_audio_component_copy_name(
    _ rawComponent: UnsafeMutableRawPointer?,
    _ outStatus: UnsafeMutablePointer<Int32>?
) -> UnsafeMutablePointer<CChar>? {
    guard let outStatus else {
        return nil
    }

    var cfName: Unmanaged<CFString>?
    let status = AudioComponentCopyName(audioComponent(from: rawComponent), &cfName)
    outStatus.pointee = status
    guard status == noErr, let cfName else {
        return nil
    }
    return cStringCopy(cfName.takeRetainedValue() as String)
}

@_cdecl("at_audio_component_get_description")
public func at_audio_component_get_description(
    _ rawComponent: UnsafeMutableRawPointer?,
    _ outDescription: UnsafeMutablePointer<AudioComponentDescription>?
) -> Int32 {
    guard let outDescription else {
        return Int32(kAudio_ParamError)
    }
    return AudioComponentGetDescription(audioComponent(from: rawComponent), outDescription)
}

@_cdecl("at_audio_component_get_version")
public func at_audio_component_get_version(
    _ rawComponent: UnsafeMutableRawPointer?,
    _ outVersion: UnsafeMutablePointer<UInt32>?
) -> Int32 {
    guard let outVersion else {
        return Int32(kAudio_ParamError)
    }
    return AudioComponentGetVersion(audioComponent(from: rawComponent), outVersion)
}

@_cdecl("at_audio_component_instance_new")
public func at_audio_component_instance_new(
    _ rawComponent: UnsafeMutableRawPointer?,
    _ outHandle: UnsafeMutablePointer<UnsafeMutableRawPointer?>?
) -> Int32 {
    guard let outHandle else {
        return Int32(kAudio_ParamError)
    }

    var instance: AudioComponentInstance?
    let status = AudioComponentInstanceNew(audioComponent(from: rawComponent), &instance)
    if status == noErr, let instance {
        outHandle.pointee = retainObject(AudioComponentInstanceBox(instance))
    } else {
        outHandle.pointee = nil
    }
    return status
}

@_cdecl("at_audio_component_instance_raw")
public func at_audio_component_instance_raw(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let handle else {
        return nil
    }
    let box: AudioComponentInstanceBox = takeUnretained(handle)
    return toRawPointer(box.value)
}

@_cdecl("at_audio_component_instance_release")
public func at_audio_component_instance_release(_ handle: UnsafeMutableRawPointer?) {
    guard let handle else {
        return
    }
    releaseObject(handle, as: AudioComponentInstanceBox.self)
}

@_cdecl("at_audio_component_instance_get_component")
public func at_audio_component_instance_get_component(
    _ rawInstance: UnsafeMutableRawPointer?
) -> UnsafeMutableRawPointer? {
    let component = AudioComponentInstanceGetComponent(audioComponentInstance(from: rawInstance))
    return retainObject(AudioComponentBox(component))
}

File: src/audio_component.rs
use crate::{
    ffi,
    internal::{status_to_result, string_from_owned_ptr},
    AudioComponentDescription, AudioComponentInstanceRef, AudioComponentRef,
    AudioComponentValidationResult, AudioToolboxError, CFDictionaryRef, Result,
};
use std::{mem::MaybeUninit, ptr::NonNull};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
/// Wrapper around an AudioToolbox.framework `AudioComponent`.
pub struct AudioComponent(NonNull<std::ffi::c_void>);

#[derive(Debug, Clone)]
/// Iterator over components returned by `AudioComponentFindNext`.
pub struct AudioComponentIter {
    previous: Option<AudioComponent>,
    description: AudioComponentDescription,
}

#[derive(Debug)]
/// Owning wrapper around an AudioToolbox.framework `AudioComponentInstance`.
pub struct AudioComponentInstance {
    handle: *mut std::ffi::c_void,
    raw: AudioComponentInstanceRef,
}

impl AudioComponent {
    /// Wraps `AudioComponentCount`.
    pub fn count(description: AudioComponentDescription) -> u32 {
        unsafe { ffi::audio_component::at_audio_component_count(std::ptr::from_ref(&description)) }
    }

    /// Wraps `AudioComponentFindNext`.
    pub fn find_next(
        previous: Option<Self>,
        description: AudioComponentDescription,
    ) -> Option<Self> {
        let previous_raw = previous.map_or(std::ptr::null_mut(), |component| component.as_raw());
        let handle = unsafe {
            ffi::audio_component::at_audio_component_find_next(
                previous_raw.cast(),
                std::ptr::from_ref(&description),
            )
        };
        let raw = NonNull::new(unsafe { ffi::audio_component::at_audio_component_raw(handle) })?;
        unsafe { ffi::audio_component::at_audio_component_release(handle) };
        Some(Self(raw))
    }

    /// Returns an iterator that walks `AudioComponentFindNext`.
    pub fn iter(description: AudioComponentDescription) -> AudioComponentIter {
        AudioComponentIter {
            previous: None,
            description,
        }
    }

    /// Returns the wrapped `AudioComponentRef`.
    pub fn as_raw(&self) -> AudioComponentRef {
        self.0.as_ptr()
    }

    /// Wraps `AudioComponentCopyName`.
    pub fn copy_name(&self) -> Result<String> {
        let mut status = 0;
        let ptr = unsafe {
            ffi::audio_component::at_audio_component_copy_name(self.as_raw().cast(), &mut status)
        };
        status_to_result("AudioComponentCopyName", status)?;
        string_from_owned_ptr("AudioComponentCopyName", ptr)
    }

    /// Wraps `AudioComponentGetDescription`.
    pub fn description(&self) -> Result<AudioComponentDescription> {
        let mut description = MaybeUninit::uninit();
        let status = unsafe {
            ffi::audio_component::at_audio_component_get_description(
                self.as_raw().cast(),
                description.as_mut_ptr(),
            )
        };
        status_to_result("AudioComponentGetDescription", status)?;
        Ok(unsafe { description.assume_init() })
    }

    /// Wraps `AudioComponentGetVersion`.
    pub fn version(&self) -> Result<u32> {
        let mut version = 0_u32;
        let status = unsafe {
            ffi::audio_component::at_audio_component_get_version(self.as_raw().cast(), &mut version)
        };
        status_to_result("AudioComponentGetVersion", status)?;
        Ok(version)
    }

    /// Wraps `AudioComponentCopyConfigurationInfo`.
    pub fn copy_configuration_info_raw(&self) -> Result<CFDictionaryRef> {
        let mut configuration_info = std::ptr::null();
        let status = unsafe {
            ffi::audio_component::at_audio_component_copy_configuration_info(
                self.as_raw(),
                &mut configuration_info,
            )
        };
        status_to_result("AudioComponentCopyConfigurationInfo", status)?;
        Ok(configuration_info)
    }

    /// Wraps `AudioComponentValidate`.
    pub fn validate_raw(
        &self,
        validation_parameters: Option<CFDictionaryRef>,
    ) -> Result<AudioComponentValidationResult> {
        let mut validation_result = 0_u32;
        let status = unsafe {
            ffi::audio_component::at_audio_component_validate(
                self.as_raw(),
                validation_parameters.unwrap_or(std::ptr::null()),
                &mut validation_result,
            )
        };
        status_to_result("AudioComponentValidate", status)?;
        Ok(validation_result)
    }

    /// Wraps `AudioComponentInstanceNew`.
    ///
    /// The returned wrapper owns the underlying AudioToolbox.framework handle and releases it on drop.
    pub fn new_instance(&self) -> Result<AudioComponentInstance> {
        let mut handle = std::ptr::null_mut();
        let status = unsafe {
            ffi::audio_component::at_audio_component_instance_new(self.as_raw().cast(), &mut handle)
        };
        status_to_result("AudioComponentInstanceNew", status)?;
        let raw: AudioComponentInstanceRef =
            unsafe { ffi::audio_component::at_audio_component_instance_raw(handle) }.cast();
        if raw.is_null() {
            return Err(AudioToolboxError::message(
                "AudioComponentInstanceNew",
                "framework returned a null AudioComponentInstance",
            ));
        }
        Ok(AudioComponentInstance { handle, raw })
    }
}

impl Iterator for AudioComponentIter {
    type Item = AudioComponent;

    fn next(&mut self) -> Option<Self::Item> {
        let next = AudioComponent::find_next(self.previous, self.description);
        self.previous = next;
        next
    }
}

impl AudioComponentInstance {
    /// Returns the wrapped `AudioComponentInstanceRef`.
    pub fn as_raw(&self) -> AudioComponentInstanceRef {
        self.raw
    }

    /// Wraps `AudioComponentInstanceGetComponent`.
    pub fn component(&self) -> Result<AudioComponent> {
        let handle = unsafe {
            ffi::audio_component::at_audio_component_instance_get_component(self.raw.cast())
        };
        let raw = NonNull::new(unsafe { ffi::audio_component::at_audio_component_raw(handle) })
            .ok_or_else(|| {
                AudioToolboxError::message(
                    "AudioComponentInstanceGetComponent",
                    "framework returned a null component",
                )
            })?;
        unsafe { ffi::audio_component::at_audio_component_release(handle) };
        Ok(AudioComponent(raw))
    }

    /// Wraps `AudioComponentInstanceCanDo`.
    pub fn can_do(&self, selector_id: i16) -> bool {
        unsafe {
            ffi::audio_component::at_audio_component_instance_can_do(self.raw, selector_id) != 0
        }
    }

    /// Wraps `AudioComponentInstanceDispose`.
    pub fn dispose(mut self) -> Result<()> {
        self.release();
        Ok(())
    }

    fn release(&mut self) {
        if !self.handle.is_null() {
            unsafe { ffi::audio_component::at_audio_component_instance_release(self.handle) };
            self.handle = std::ptr::null_mut();
            self.raw = std::ptr::null_mut();
        }
    }
}

impl Drop for AudioComponentInstance {
    fn drop(&mut self) {
        self.release();
    }
}
